File: 2023-12-18/leetcode_332.py
from collections import defaultdict
class Solution:
    def findItinerary(self, tickets):
        graph = defaultdict(list)
        route = []

        for departure, arrival in sorted(tickets):
            graph[departure].append(arrival)

        def dfs(start = "JFK"):
            while graph[start]:
                dfs(graph[start].pop(0))
            route.append(start)
        dfs()
        return route[::-1]

File: 2023-12-18/leetcode_332.js
const findItinerary = (tickets) => {
  const graph = new Map();
  route = [];

  tickets.sort().forEach(([departure, arrival]) => {
    if (!graph[departure]) {
      graph[departure] = [];
    }
    graph[departure].push(arrival);
  });

  const dfs = (start = 'JFK') => {
    while (start in graph && !!graph[start].length) {
      dfs(graph[start].shift());
    }
    route.push(start);
  };
  dfs();
  return route.reverse();
};
